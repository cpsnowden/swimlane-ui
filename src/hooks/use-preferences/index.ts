export { usePreferences } from "./use-preferences.hook";
