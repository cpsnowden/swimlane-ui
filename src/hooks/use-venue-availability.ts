import { useQuery } from "@tanstack/react-query";
import dayjs from "dayjs";

import { api } from "src/api";
import { Session } from "src/types";

interface LaneTimeDTO {
  startAtUTC: string;
  endAtUTC: string;
  laneName: string;
  spaces: number;
}

const getVenueTimes = async (
  operator: string,
  venueId: string,
  date: string
): Promise<Array<Session>> => {
  const response = await api.get<Array<LaneTimeDTO>>(
    `/operators/${operator}/venues/${venueId}/times/${date}`
  );
  return response.data.map((session) => {
    return {
      startAt: dayjs(session.startAtUTC),
      endAt: dayjs(session.endAtUTC),
      laneName: session.laneName,
      spaces: session.spaces,
    };
  });
};

interface UseGetVenueAvailabilityReturn {
  sessions: Session[];
  isSessionsLoading: boolean;
}

export const useGetVenueAvailability = (
  operator: string,
  venueId: string,
  date: string
): UseGetVenueAvailabilityReturn => {
  const { data: sessions, isLoading: isSessionsLoading } = useQuery({
    queryKey: ["venue-availability", operator, venueId, date],
    queryFn: () => getVenueTimes(operator, venueId, date),
  });
  return { sessions: sessions ?? [], isSessionsLoading };
};
