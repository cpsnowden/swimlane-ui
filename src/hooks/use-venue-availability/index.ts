export { useVenueAvailability } from "./use-venue-availability.hook";
