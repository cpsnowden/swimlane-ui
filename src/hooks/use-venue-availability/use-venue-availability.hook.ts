import { useQuery } from "@tanstack/react-query";
import dayjs from "dayjs";

import { api } from "src/lib/axios";
import { Session } from "src/types";

interface LaneTimeDTO {
  startAtUTC: string;
  endAtUTC: string;
  laneName: string;
  spaces: number;
}

const getVenueTimes = async (
  operator: string,
  venueId: string,
  date: string
): Promise<Array<Session>> => {
  const response = await api.get<Array<LaneTimeDTO>>(
    `/operators/${operator}/venues/${venueId}/times/${date}`
  );
  return response.data.map((session) => {
    return {
      startAt: dayjs(session.startAtUTC),
      endAt: dayjs(session.endAtUTC),
      laneName: session.laneName,
      spaces: session.spaces,
    };
  });
};

interface UseVenueAvailabilityReturn {
  sessions: Session[];
  isSessionsLoading: boolean;
}

export const useVenueAvailability = (
  operator: string,
  venueId: string,
  date: string
): UseVenueAvailabilityReturn => {
  const { data: sessions, isLoading: isSessionsLoading } = useQuery({
    queryKey: ["venue-availability", operator, venueId, date],
    queryFn: () => getVenueTimes(operator, venueId, date),
  });
  return { sessions: sessions ?? [], isSessionsLoading };
};
