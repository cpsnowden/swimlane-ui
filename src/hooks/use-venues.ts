import { useQuery } from "@tanstack/react-query";

import { api } from "src/api";
import { Venue } from "src/types";

interface VenueDTO {
  name: string;
  slug: string;
  operator: string;
}

const getVenues = async (): Promise<Array<Venue>> => {
  const response = await api.get<Array<VenueDTO>>("venues");
  return response.data;
};

interface UseGetVenuesReturn {
  venues: Venue[];
  isVenuesLoading: boolean;
}

export const useGetVenues = (): UseGetVenuesReturn => {
  const { data: venues, isLoading: isVenuesLoading } = useQuery({
    queryKey: ["venues"],
    queryFn: getVenues,
  });
  return { venues: venues ?? [], isVenuesLoading };
};
