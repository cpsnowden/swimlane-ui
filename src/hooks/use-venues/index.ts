export { useVenues } from "./use-venues.hook";
