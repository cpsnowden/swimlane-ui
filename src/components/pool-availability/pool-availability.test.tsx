import { render, screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import dayjs from "dayjs";
import { vi } from "vitest";

import { useGetVenueAvailability } from "src/hooks/use-venue-availability";

import { PoolAvailability } from ".";

const mockUseGetVenueAvailability =
  useGetVenueAvailability as jest.MockedFunction<
    typeof useGetVenueAvailability
  >;

vi.mock("src/hooks/use-venue-availability", () => {
  return {
    useGetVenueAvailability: vi.fn(),
  };
});

describe("PoolAvailability", () => {
  it("should render loading indicator whilst loading", () => {
    mockUseGetVenueAvailability.mockReturnValue({
      sessions: [],
      isSessionsLoading: true,
    });

    const date = dayjs("2023-05-21");

    render(
      <PoolAvailability
        operator={"operator-id"}
        venueId={"venue-id"}
        venueName={"venue-name"}
        date={date}
        onRemoveClick={vi.fn()}
      />
    );

    expect(screen.getByText("Loading Availability")).toBeTruthy();
    expect(mockUseGetVenueAvailability).toHaveBeenCalledWith(
      "operator-id",
      "venue-id",
      "2023-05-21"
    );
  });

  it("should render list of sessions", () => {
    mockUseGetVenueAvailability.mockReturnValue({
      sessions: [
        {
          startAt: dayjs("2023-05-21T08:00:00"),
          endAt: dayjs("2023-05-21T08:59:00"),
          laneName: "Lane 1",
          spaces: 1,
        },
        {
          startAt: dayjs("2023-05-21T09:00:00"),
          endAt: dayjs("2023-05-21T09:59:00"),
          laneName: "Lane 2",
          spaces: 2,
        },
      ],
      isSessionsLoading: false,
    });

    const date = dayjs("2023-05-21");

    render(
      <PoolAvailability
        operator={"operator-id"}
        venueId={"venue-id"}
        venueName={"venue-name"}
        date={date}
        onRemoveClick={vi.fn()}
      />
    );

    const sessionList = screen.getByRole("list", { name: /sessions/i });
    const sessions = within(sessionList).getAllByRole("listitem");

    // Header + 2
    expect(sessions).toHaveLength(3);
    expect(sessions[1]).toMatchSnapshot();
  });

  it("should render no results if no sessions", () => {
    mockUseGetVenueAvailability.mockReturnValue({
      sessions: [],
      isSessionsLoading: false,
    });

    render(
      <PoolAvailability
        operator={"operator-id"}
        venueId={"venue-id"}
        venueName={"venue-name"}
        date={dayjs("2023-05-21")}
        onRemoveClick={vi.fn()}
      />
    );

    expect(screen.getByText("No results")).toBeTruthy();
  });

  it("should render error alert when query fails", () => {
    //TODO
  });

  it("should call onRemoveClick", async () => {
    mockUseGetVenueAvailability.mockReturnValue({
      sessions: [],
      isSessionsLoading: false,
    });

    const onRemoveClickMock = vi.fn();

    render(
      <PoolAvailability
        operator={"operator-id"}
        venueId={"venue-id"}
        venueName={"venue-name"}
        date={dayjs("2023-05-21")}
        onRemoveClick={onRemoveClickMock}
      />
    );

    const closeButton = screen.getByTestId("CloseIcon");
    userEvent.click(closeButton);

    await waitFor(() => {
      expect(onRemoveClickMock).toHaveBeenCalled();
    });
  });
});
