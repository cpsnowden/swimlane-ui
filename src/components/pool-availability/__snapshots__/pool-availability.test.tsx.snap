// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PoolAvailability > should render list of sessions 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: 100%;
  box-sizing: border-box;
  text-align: left;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 16px;
  padding-right: 16px;
}

.emotion-0.Mui-focusVisible {
  background-color: rgba(0, 0, 0, 0.12);
}

.emotion-0.Mui-selected {
  background-color: rgba(25, 118, 210, 0.08);
}

.emotion-0.Mui-selected.Mui-focusVisible {
  background-color: rgba(25, 118, 210, 0.2);
}

.emotion-0.Mui-disabled {
  opacity: 0.38;
}

.emotion-1 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  min-width: 0;
  margin-top: 6px;
  margin-bottom: 6px;
}

.emotion-2 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  display: block;
}

.emotion-3 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.43;
  letter-spacing: 0.01071em;
  color: rgba(0, 0, 0, 0.6);
  display: block;
}

<li
  class="MuiListItem-root MuiListItem-gutters MuiListItem-padding emotion-0"
>
  <div
    class="MuiListItemText-root MuiListItemText-multiline emotion-1"
  >
    <span
      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary emotion-2"
    >
      08:00 - 08:59 (1 spaces)
    </span>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary emotion-3"
    >
      Lane 1
    </p>
  </div>
</li>
`;
